openquake/hazardlib/gsim/sgobba_2020.py

```diff
@@ -0,0 +1,299 @@
+        corresponds to the cluster id (available cluster indexes are 1, 4
+        and 5), the corresponding correction id applied.
+    """
+
+    #: Supported tectonic region type is 'active shallow crust'
+    DEFINED_FOR_TECTONIC_REGION_TYPE = const.TRT.ACTIVE_SHALLOW_CRUST
+
+    #: Set of :mod:`intensity measure types <openquake.hazardlib.imt>`
+    #: this GSIM can calculate. A set should contain classes from module
+    #: :mod:`openquake.hazardlib.imt`.
+    DEFINED_FOR_INTENSITY_MEASURE_TYPES = {
+        PGA,
+        SA
+    }
+
+    #: Supported intensity measure component is the geometric mean of two
+    #: horizontal components
+    DEFINED_FOR_INTENSITY_MEASURE_COMPONENT = const.IMC.AVERAGE_HORIZONTAL
+
+    #: Supported standard deviation types are inter-event, intra-event
+    #: and total
+    DEFINED_FOR_STANDARD_DEVIATION_TYPES = {
+        const.StdDev.TOTAL,
+        const.StdDev.INTER_EVENT,
+        const.StdDev.INTRA_EVENT
+    }
+
+    #: Required site parameter is not set
+    REQUIRES_SITES_PARAMETERS = set()
+
+    #: Required rupture parameters is magnitude, ev_lat, ev_lon
+    REQUIRES_RUPTURE_PARAMETERS = {'mag', 'ev_lat', 'ev_lon'}
+
+    #: Required distance measure is Rjb
+    REQUIRES_DISTANCES = {'rjb'}
+
+    def __init__(self, event_id=None, directionality=False, cluster=None,
+                 site=False, bedrock=False, **kwargs):
+        super().__init__(event_id=event_id,
+                         directionality=directionality,
+                         cluster=cluster,
+                         **kwargs)
+        self.event_id = event_id
+        self.directionality = directionality
+        self.cluster = cluster
+        self.site = site
+        self.bedrock = bedrock
+
+        # Load the table with the between event coefficients
+        if event_id is not None:
+            self.event_id = event_id
+            fname = os.path.join(DATA_FOLDER, 'event.csv')
+            # Create dataframe with events
+            df = pd.read_csv(fname, dtype={'id': str})
+            df.set_index('id', inplace=True)
+            self.df = df
+            assert event_id in df.index.values
+
+    def get_mean_and_stddevs(self, sites, rup, dists, imt, stddev_types):
+        """
+        Eq.1 - page 2
+        """
+        # Ergodic coeffs
+        C = self.COEFFS[imt]
+        # between-event
+        if self.event_id is not None:
+            label = "dBe_{:s}".format(imt.__str__())
+            self.be = self.df.loc[self.event_id][label]
+            self.be_std = 0.0
+        else:
+            self.be_std = C['tau_ev']
+            self.be = 0.0
+
+        # Get site indexes. They are used for the site correction and the
+        # cluster (path) correction
+        # Load the coordinate of the grid
+
+        fname = os.path.join(DATA_FOLDER, 'grid.csv')
+        coo = np.fliplr(np.loadtxt(fname, delimiter=";"))
+        # Create a spatial index
+        kdt = cKDTree(coo)
```

Comment:
If the grid is fixed then wouldn't it be better to create the KDTree upon instantiation of the class rather than in `get_mean_and_stddevs`? As this function is called potentially millions of times then I can't see this being functional if on each call it requires the import of a grid and construction of a KDTree.

Reply:
I have moved that block in the init.